// @dialectlabs/blockchain-sdk-aptos/lib/cjs/auth/ed25519/aptos-ed25519-token-signer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DialectWalletAdapterAptosEd25519TokenSigner = exports.AptosEd25519TokenSigner = exports.APTOS_ED25519_TOKEN_SIGNER_ALG = void 0;
const tslib_1 = require("tslib");
const aptos_1 = require("aptos");
const aptos_public_key_1 = require("../aptos-public-key");
const constants_1 = require("../../sdk/constants");
exports.APTOS_ED25519_TOKEN_SIGNER_ALG = `${constants_1.DIALECT_BLOCKCHAIN_SDK_TYPE_APTOS}-ed25519`;
class AptosEd25519TokenSigner {
    constructor() {
        this.alg = exports.APTOS_ED25519_TOKEN_SIGNER_ALG;
    }
}
exports.AptosEd25519TokenSigner = AptosEd25519TokenSigner;
class DialectWalletAdapterAptosEd25519TokenSigner extends AptosEd25519TokenSigner {
    constructor(dialectWalletAdapter) {
        super();
        this.dialectWalletAdapter = dialectWalletAdapter;
    }
    get subject() {
        const address = this.dialectWalletAdapter.address;
        const hexString = aptos_1.HexString.ensure(address.toString());
        return hexString.toString();
    }
    get subjectPublicKey() {
        const publicKey = this.dialectWalletAdapter.publicKey;
        const hexString = aptos_1.HexString.ensure(publicKey.toString());
        return new aptos_public_key_1.AptosPubKey(hexString);
    }
    sign(payload) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const stringPayload = new TextDecoder().decode(payload);
            const signatureString = yield this.dialectWalletAdapter.signMessage(stringPayload);
            const hexString = aptos_1.HexString.ensure(signatureString);
            const signature = hexString.toUint8Array();
            return {
                payload,
                signature,
            };
        });
    }
}
exports.DialectWalletAdapterAptosEd25519TokenSigner = DialectWalletAdapterAptosEd25519TokenSigner;
